### PR TITLE
Feature: BaggingClassifier

### PR DESCRIPTION
**Pull request by OlegStotsky**

### CODE REVIEW

File: src/lib/ensemble/bagging.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+ *  baseEstimator: LogisticRegression,
+ *  maxSamples: 1.0,
+ * });
+ * const X = [[1], [2], [3], [4], [5]];
+ * const y = [1, 1, 1, 1, 1];
+ * classifier.fit(X, y);
+ * classifier.predict(X);
+ */
+export class BaggingClassifier {
+  private BaseEstimator: any;
```
Comment (JasonShin) [inline src/lib/ensemble/bagging.ts]: I think we can correctly type this later on. We have `IMlModel` interface but it's not really useful right now. I can work on this.

File: src/lib/ensemble/bagging.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+   *  Is used in conjunction with @param maxSamplesIsFloating.
+   *  If @param maxSamplesIsFloating is false, then draw maxSamples samples.
+   *  If @param maxSamplesIsFloating is true, then draw max_samples * shape(X)[0] samples.
+   * @param maxFeatures - The number of features to draw from X to train each base estimator.
+   *  Is used in conjunction with @param maxFeaturesIsFloating
+   *  If @param maxFeaturesIsFloating is false, then draw max_features features.
+   *  If @param maxFeaturesIsFloating is true, then draw max_features * shape(X)[1] features.
+   * @param bootstrapSamples - Whether samples are drawn with replacement. If false, sampling without replacement is performed.
+   * @param bootstrapFeatures - Whether features are drawn with replacement.
+   * @param estimatorOptions - constructor options for BaseEstimator.
```
Comment (JasonShin) [inline src/lib/ensemble/bagging.ts]: Missing docs for `maxSamplesIsFloating` and `maxFeaturesIsFloating`

Comment (OlegStotsky) [reply src/lib/ensemble/bagging.ts]: Their meaning is described in maxSamples and maxFeatures

File: src/lib/utils/MathExtra.ts [hunk 3]
```diff
@@ -316,9 +318,110 @@ const inner = (a, b) => {
...
+ *  If @param maxSamplesIsFloating is false, then draw maxSamples samples.
+ *  If @param maxSamplesIsFloating is true, then draw max_samples * shape(X)[0] samples.
+ * @param maxFeatures - The number of features to draw from X to train each base estimator.
+ *  Is used in conjunction with @param maxFeaturesIsFloating
+ *  If @param maxFeaturesIsFloating is false, then draw max_features features.
+ *  If @param maxFeaturesIsFloating is true, then draw max_features * shape(X)[1] features.
+ * @param bootstrapSamples - Whether samples are drawn with replacement. If false, sampling without replacement is performed.
+ * @param bootstrapFeatures - Whether features are drawn with replacement.
+ */
+const generateRandomSubsetOfMatrix = <T>(
```
Comment (JasonShin) [inline src/lib/utils/MathExtra.ts]: Ok, this guy is causing the issue when generating a doc, let me fix this

**Comment by JasonShin** [comment]
amazing work

File: src/lib/utils/MathExtra.ts (hunk 3 above)
Comment (JasonShin) [reply src/lib/utils/MathExtra.ts]: resolved by ignoring it

File: src/lib/ensemble/bagging.ts (hunk 2 above)
Comment (JasonShin) [reply src/lib/ensemble/bagging.ts]: Ok, let me pull them out so it can be picked up by the doc generator

File: src/lib/ensemble/bagging.ts (hunk 1 above)
Comment (JasonShin) [reply src/lib/ensemble/bagging.ts]: Opened an issue for this https://github.com/machinelearnjs/machinelearnjs/issues/245

File: src/lib/ensemble/bagging.ts (hunk 2 above)
Comment (JasonShin) [reply src/lib/ensemble/bagging.ts]: Ok resolved

**Comment by JasonShin** [comment]
@OlegStotsky congratulations Oleg! This is an amazing addition to machinelearn.js and I learned something new!

**Pull request merged**